### PR TITLE
#162633507 Add Favorite articles Functionality

### PR DESCRIPTION
**Pull request by NabunyaLilian**
**What does this PR do?**

Enables users to favorite articles of interest.
**Description of Tasks completed**

* Create a model for favoriting articles and unfavoriting articles
* Create a serializer and view for the favorite articles and unfavorite articles
* Test the functionality to make sure it performs as expected.

**How to manually test**
-  Run `python manage.p runserver`
-  Run `python manage.py test`
- Register and login
- Create an article
- Register and login someone else
- Someone else accesses favorite endpoint to favorite your article.
- Someone can also unfavorite your artice

* Endpoints to be tested*
- ` http://127.0.0.1:8000/api/v1/users/register` 
-  `http://127.0.0.1:8000/api/v1/users/login/`
- `http://127.0.0.1:8000/api/v1/articles/`
-  `http://127.0.0.1:8000/api/v1/articles/<slug:slug>/favorite` Favorite Article
-  `http://127.0.0.1:8000/api/v1/articles/<slug:slug>/favorite/<int:pk>/ ` Unfavorite Article
**Relevant pivotal tracker story**

[DELIVERS #162633507](https://www.pivotaltracker.com/story/show/162633507)

### CODE REVIEW

File: authors/apps/articles/helper.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+            favourite = kwargs.get('model').objects.all().filter(
+                article=kwargs.get('article_id')
+            ) 
+            filtered_favourite = favourite.filter(author=kwargs.get('user_id'))
+            if favourite and filtered_favourite:
+                return True
+            else:
+                return False    
+        except Exception as e:
+            print(e)
```
Comment (jeanjoe) [inline authors/apps/articles/helper.py]: Please remove the `print` statement.

File: authors/apps/articles/urls.py [hunk 2]
```diff
@@ -1,11 +1,13 @@
 from django.urls import path
 
 from .views import (ListCreateArticle,
-                    RetrieveUpdateDestroyArticle, RetrieveAuthorArticles)
+                    RetrieveUpdateDestroyArticle, RetrieveAuthorArticles, FavoriteArticlesView, UnfavoriteArticleView)
```
Comment (jeanjoe) [inline authors/apps/articles/urls.py:4]: Please reduce this line to less than 80 characters

File: authors/apps/articles/urls.py [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
 
 app_name = "articles"
 urlpatterns = [
     path('', ListCreateArticle.as_view(), name="articles_list"),
     path('<slug:slug>/', RetrieveUpdateDestroyArticle.as_view(), name="article_detail"),
     path('author/<int:pk>/', RetrieveAuthorArticles.as_view(), name="author_articles"),
+    path('<slug:slug>/favorite', FavoriteArticlesView.as_view(), name="favorite-articles"),
+    path('<slug:slug>/favorite/<int:pk>/', UnfavoriteArticleView.as_view(), name="unfavorite-articles")
```
Comment (jeanjoe) [inline authors/apps/articles/urls.py]: Reduce line to less than 80 characters

File: authors/apps/articles/views.py [hunk 4]
```diff
@@ -32,3 +37,48 @@ class RetrieveAuthorArticles(generics.ListAPIView):
...
+    )
+
+    def perform_create(self, serializer):
+        article = Article.objects.get(slug=self.kwargs.get('slug'))
+        if article.author.id == self.request.user.id:
+            message = "You're not authorised to favorite your own artcle"
+            raise NotAcceptable(detail=message, code=status.HTTP_406_NOT_ACCEPTABLE)
+        favor = None
+        try:
+            favor = FavoriteArticle.objects.get(author=self.request.user.id, article=article.id)
```
Comment (jeanjoe) [inline authors/apps/articles/views.py]: Same line issue

**Review by jeanjoe — CHANGES_REQUESTED** [review]

**Comment by NabunyaLilian** [comment]
Thanks @jeanjoe  for that feedback, let me work on it

**Pull request merged**